examples/rap_example.cpp

```diff
@@ -0,0 +1,112 @@
+#include <stdio.h>
+
+#include <iostream>
+
+#include "camp/camp.hpp"
+#include "umpire/ResourceManager.hpp"
+#include "umpire/Umpire.hpp"
+#include "umpire/strategy/QuickPool.hpp"
+#include "umpire/strategy/ResourceAwarePool.hpp"
+
+using namespace camp::resources;
+
+#if defined(UMPIRE_ENABLE_CUDA)
+using resource_type = Cuda;
+#elif defined(UMPIRE_ENABLE_HIP)
+using resource_type = Hip;
+#else
+using resource_type = Host;
+#endif
+
+constexpr int NUM_THREADS = 64;
+
+void host_sleep(double* ptr)
+{
+  double i = 0.0;
+  while (i < 1000000) {
+    double y = i;
+    y++;
+    i = y;
+  }
+  *ptr = i;
+  ptr++;
+}
+
+#if defined(UMPIRE_ENABLE_CUDA) || defined(UMPIRE_ENABLE_HIP)
+constexpr int BLOCK_SIZE = 16;
+using clock_value_t = long long;
+
+__device__ clock_value_t my_clock()
+{
+  return clock64();
+}
+
+__device__ void sleep(clock_value_t sleep_cycles)
+{
+  clock_value_t start = my_clock();
+  clock_value_t cycles_elapsed;
+  do {
+    cycles_elapsed = my_clock() - start;
+  } while (cycles_elapsed < sleep_cycles);
+}
+
+__global__ void do_sleep()
+{
+  // sleep - works still at 1000, so keeping it at 100k
+  sleep(10000000);
+}
+#endif
+
+int main(int, char**)
+{
+  auto& rm = umpire::ResourceManager::getInstance();
+
+#if defined(UMPIRE_ENABLE_CUDA) || defined(UMPIRE_ENABLE_HIP)
+  auto pool = rm.makeAllocator<umpire::strategy::ResourceAwarePool>("rap-pool", rm.getAllocator("UM"));
+  const int NUM_BLOCKS = NUM_THREADS / BLOCK_SIZE;
+#else
+  auto pool = rm.makeAllocator<umpire::strategy::ResourceAwarePool>("rap-pool", rm.getAllocator("HOST"));
+#endif
+
+  // Create camp resources for RAP
+  resource_type d1, d2;
+  Resource r1{d1}, r2{d2};
+
+  // allocate memory in the pool with r1
+  double* a = static_cast<double*>(pool.allocate(r1, NUM_THREADS * sizeof(double)));
```

Comment:
Is it legal to pass in d1 directly instead of r1?

Reply:
This dumbfounded me - if I remove all instances of r1 and r2 and replace with d1 and d2 only, everything still works. I did not expect that to work.... so I need to ask about this and then potentially change some documentation I have... Thanks for pointing this out!

Reply:
When I test: `EXPECT_EQ(getResource(m_pool, ptr), r1);` - That check only works with r1 or r2, not d1 or d2. I thnk this is because I store the generic resource in `m_resource` of the pool chunk. So whenever a user calls `getResource` the user should expect the generic resource (r1) only to be returned, not the concrete one (d1).

It would have been nice to just only use d1/d2 or only use r1/r2, but it seems like there is no way to do that. I can't get around a kernel launch without using d1.get_stream() (so I need d1, can't use r1) ... but then i can't get around the return of getResource which will only use r1, can't expect d1 ... It looks like there has to be this confusing mix of generic (r1) and concrete (d1) resource. Is that correct?

Reply:
This comes from the issues we had enabling comparison operators between generic and typed resources. You could comparethem by explicitly casting the concrete to generic `EXPECT_EQ(getResource(m_pool, ptr), Resource(d1));`